Vineyard fails to work with latest pandas (v1.3.0)

<!--
Thanks for your contribution! please review https://github.com/v6d-io/v6d/blob/main/CONTRIBUTING.rst before opening an issue.
-->

Describe your problem
---------------------

The changes in pandas breaks vineyard's builder & resolver,

```python
obj = Object <"o00000350623e16cf": vineyard::DataFrame>
resolver = <vineyard.core.resolver.ResolverContext object at 0x7f1fb99ea370>

    def pandas_dataframe_resolver(obj, resolver):
        meta = obj.meta
        columns = from_json(meta['columns_'])
        if not columns:
            return pd.DataFrame()
        # ensure zero-copy
        blocks = []
        index_size = 0
        for idx, name in enumerate(columns):
            np_value = resolver.run(obj.member('__values_-value-%d' % idx))
            index_size = len(np_value)
            # ndim: 1 for SingleBlockManager/Series, 2 for BlockManager/DataFrame
>           blocks.append(Block(np.expand_dims(np_value, 0), slice(idx, idx + 1, 1), ndim=2))
E           TypeError: Block() takes no arguments
```

See also: https://github.com/v6d-io/v6d/runs/2986447125?check_suite_focus=true

Additional context
------------------

Add any other context about the problem here.

--- a/python/vineyard/data/dataframe.py
+++ b/python/vineyard/data/dataframe.py
@@ -18,7 +18,12 @@
 
 import numpy as np
 import pandas as pd
-from pandas.core.internals.blocks import Block
+try:
+    from pandas.core.internals.blocks import BlockPlacement, NumpyBlock as Block
+except:
+    BlockPlacement = None
+    NDArrayBacked = None
+    from pandas.core.internals.blocks import Block
 from pandas.core.internals.managers import BlockManager
 
 from vineyard._C import ObjectMeta
@@ -54,12 +59,17 @@ def pandas_dataframe_resolver(obj, resolver):
         np_value = resolver.run(obj.member('__values_-value-%d' % idx))
         index_size = len(np_value)
         # ndim: 1 for SingleBlockManager/Series, 2 for BlockManager/DataFrame
-        blocks.append(Block(np.expand_dims(np_value, 0), slice(idx, idx + 1, 1), ndim=2))
+        if BlockPlacement:
+            palcement = BlockPlacement(slice(idx, idx + 1, 1))
+        else:
+            palcement = slice(idx, idx + 1, 1)
+        values = np.expand_dims(np_value, 0)
+        blocks.append(Block(values, palcement, ndim=2))
     if 'index_' in meta:
         index = resolver.run(obj.member('index_'))
     else:
         index = np.arange(index_size)
-    return pd.DataFrame(BlockManager(blocks, [columns, index]))
+    return pd.DataFrame(BlockManager(blocks, [pd.Index(columns), index]))
 
 
 def pandas_sparse_array_builder(client, value, builder, **kw):

--- a/python/vineyard/data/dataframe.py
+++ b/python/vineyard/data/dataframe.py
@@ -22,7 +22,6 @@ try:
     from pandas.core.internals.blocks import BlockPlacement, NumpyBlock as Block
 except:
     BlockPlacement = None
-    NDArrayBacked = None
     from pandas.core.internals.blocks import Block
 from pandas.core.internals.managers import BlockManager
 
@@ -60,11 +59,11 @@ def pandas_dataframe_resolver(obj, resolver):
         index_size = len(np_value)
         # ndim: 1 for SingleBlockManager/Series, 2 for BlockManager/DataFrame
         if BlockPlacement:
-            palcement = BlockPlacement(slice(idx, idx + 1, 1))
+            placement = BlockPlacement(slice(idx, idx + 1, 1))
         else:
-            palcement = slice(idx, idx + 1, 1)
+            placement = slice(idx, idx + 1, 1)
         values = np.expand_dims(np_value, 0)
-        blocks.append(Block(values, palcement, ndim=2))
+        blocks.append(Block(values, placement, ndim=2))
     if 'index_' in meta:
         index = resolver.run(obj.member('index_'))
     else:

--- a/python/vineyard/data/series.py
+++ b/python/vineyard/data/series.py
@@ -19,7 +19,11 @@
 import json
 import numpy as np
 import pandas as pd
-from pandas.core.internals.blocks import Block
+try:
+    from pandas.core.internals.blocks import BlockPlacement, NumpyBlock as Block
+except:
+    BlockPlacement = None
+    from pandas.core.internals.blocks import Block
 from pandas.core.internals.managers import SingleBlockManager
 
 from vineyard._C import ObjectMeta
@@ -40,7 +44,11 @@ def pandas_series_resolver(obj, resolver):
     name = from_json(meta['name'])
     index = resolver.run(obj.member('index_'))
     np_value = resolver.run(obj.member('value_'))
-    block = Block(np_value, slice(0, len(np_value), 1), ndim=1)
+    if BlockPlacement:
+        placement = BlockPlacement(slice(0, len(np_value), 1))
+    else:
+        placement = slice(0, len(np_value), 1)
+    block = Block(np_value, placement, ndim=1)
     return pd.Series(SingleBlockManager(block, index), name=name)
 
 

--- a/python/vineyard/data/tensor.py
+++ b/python/vineyard/data/tensor.py
@@ -30,10 +30,20 @@ if pickle.HIGHEST_PROTOCOL < 5:
     import pickle5 as pickle
 
 from vineyard._C import ObjectMeta
-from .utils import from_json, to_json, build_numpy_buffer, normalize_dtype, normalize_cpptype, wrap_type
+from .utils import from_json, to_json, build_numpy_buffer, normalize_dtype, normalize_cpptype
+
 
 # Enable dynamic attribute on numpy.ndarray.
-ndarray_wrapped = wrap_type(np.ndarray)
+class ndarray(np.ndarray):
+    pass
+
+
+# n.b. don't wrap `__module__` otherwise `pickle` on such values will fail.
+#
+# ndarray.__module__ = np.ndarray.__module__
+ndarray.__name__ = np.ndarray.__name__
+ndarray.__qualname__ = np.ndarray.__qualname__
+ndarray.__doc__ = np.ndarray.__doc__
 
 
 def numpy_ndarray_builder(client, value, **kw):
@@ -67,7 +77,7 @@ def numpy_ndarray_resolver(obj):
     c_array = np.frombuffer(memoryview(obj.member('buffer_')), dtype=value_type).reshape(shape)
     # TODO: revise the memory copy of asfortranarray
     array = (c_array if order == 'C' else np.asfortranarray(c_array))
-    return array.view(ndarray_wrapped)
+    return array.view(ndarray)
 
 
 def bsr_matrix_builder(client, value, builder, **kw):

--- a/python/vineyard/data/tests/test_pickle.py
+++ b/python/vineyard/data/tests/test_pickle.py
@@ -204,3 +204,10 @@ def test_bytes_io_pandas_dataframe_mixed_columns(vineyard_client):
     object_id = vineyard_client.put(df)
     target = read_and_build(b1m, vineyard_client.get(object_id))
     pd.testing.assert_frame_equal(df, target)
+
+
+def test_bytes_io_pandas_series(vineyard_client):
+    s = pd.Series([1, 3, 5, np.nan, 6, 8], name='foo')
+    object_id = vineyard_client.put(s)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    pd.testing.assert_series_equal(s, target)

--- a/python/vineyard/data/tests/test_pickle.py
+++ b/python/vineyard/data/tests/test_pickle.py
@@ -67,8 +67,7 @@ values = [
 ]
 
 
-@pytest.mark.parametrize("block_size, value", values)
-def test_bytes_io_roundtrip(block_size, value):
+def read_and_build(block_size, value):
     reader = PickledReader(value)
     bs, nlen = [], 0
     while True:
@@ -86,7 +85,12 @@ def test_bytes_io_roundtrip(block_size, value):
     assert writer.value is None
     writer.close()
     assert writer.value is not None
-    target = writer.value
+    return writer.value
+
+
+@pytest.mark.parametrize("block_size, value", values)
+def test_bytes_io_roundtrip(block_size, value):
+    target = read_and_build(block_size, value)
 
     # compare values
     if isinstance(value, np.ndarray):
@@ -103,3 +107,100 @@ def test_bytes_io_roundtrip(block_size, value):
         pd.testing.assert_series_equal(target, value)
     else:
         assert target == value
+
+
+def test_bytes_io_numpy_ndarray(vineyard_client):
+    arr = np.random.rand(4, 5, 6)
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_allclose(arr, target)
+
+
+def test_bytes_io_empty_ndarray(vineyard_client):
+    arr = np.ones(())
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_allclose(arr, target)
+
+    arr = np.ones((0, 1))
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_allclose(arr, target)
+
+    arr = np.ones((0, 1, 2))
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_allclose(arr, target)
+
+    arr = np.ones((0, 1, 2, 3))
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_allclose(arr, target)
+
+    arr = np.zeros((), dtype='int')
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_allclose(arr, target)
+
+    arr = np.zeros((0, 1), dtype='int')
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_allclose(arr, target)
+
+    arr = np.zeros((0, 1, 2), dtype='int')
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_allclose(arr, target)
+
+    arr = np.zeros((0, 1, 2, 3), dtype='int')
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_allclose(arr, target)
+
+
+def test_bytes_io_str_ndarray(vineyard_client):
+    arr = np.array(['', 'x', 'yz', 'uvw'])
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_equal(arr, target)
+
+
+def test_object_ndarray(vineyard_client):
+    arr = np.array([1, 'x', 3.14, (1, 4)], dtype=object)
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_equal(arr, target)
+
+    arr = np.ones((), dtype='object')
+    object_id = vineyard_client.put(arr)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    np.testing.assert_equal(arr, target)
+
+
+def test_bytes_io_tensor_order(vineyard_client):
+    arr = np.asfortranarray(np.random.rand(10, 7))
+    object_id = vineyard_client.put(arr)
+    res = read_and_build(b1m, vineyard_client.get(object_id))
+    assert res.flags['C_CONTIGUOUS'] == arr.flags['C_CONTIGUOUS']
+    assert res.flags['F_CONTIGUOUS'] == arr.flags['F_CONTIGUOUS']
+
+
+def test_bytes_io_pandas_dataframe(vineyard_client):
+    df = pd.DataFrame({'a': [1, 2, 3, 4], 'b': [5, 6, 7, 8]})
+    object_id = vineyard_client.put(df)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    pd.testing.assert_frame_equal(df, target)
+
+
+def test_bytes_io_pandas_dataframe_int_columns(vineyard_client):
+    df = pd.DataFrame({1: [1, 2, 3, 4], 2: [5, 6, 7, 8]})
+    object_id = vineyard_client.put(df)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    pd.testing.assert_frame_equal(df, target)
+
+
+def test_bytes_io_pandas_dataframe_mixed_columns(vineyard_client):
+    df = pd.DataFrame({'a': [1, 2, 3, 4], 'b': [5, 6, 7, 8], 1: [9, 10, 11, 12], 2: [13, 14, 15, 16]})
+    object_id = vineyard_client.put(df)
+    target = read_and_build(b1m, vineyard_client.get(object_id))
+    pd.testing.assert_frame_equal(df, target)

--- a/python/vineyard/data/utils.py
+++ b/python/vineyard/data/utils.py
@@ -27,17 +27,6 @@ if pickle.HIGHEST_PROTOCOL < 5:
     import pickle5 as pickle
 
 
-def wrap_type(ty):
-    class T(ty):
-        pass
-
-    T.__module__ = ty.__module__
-    T.__name__ = ty.__name__
-    T.__qualname__ = ty.__qualname__
-    T.__doc__ = ty.__doc__
-    return T
-
-
 def normalize_dtype(dtype, dtype_meta=None):
     ''' Normalize a descriptive C++ type to numpy.dtype.
     '''

--- a/setup.py
+++ b/setup.py
@@ -113,7 +113,7 @@ setup(
         'numpy',
         'pandas<1.0.0; python_version<"3.6"',
         'pandas<1.2.0; python_version<"3.7"',
-        'pandas>=1.0.0,<1.3.0; python_version>="3.7"',
+        'pandas>=1.0.0; python_version>="3.7"',
         'pickle5; python_version<="3.7"',
         'pyarrow',
         'setuptools',

